File: old_implementations/limittraf.pl
#!/usr/bin/perl
###############################################################################
# limittraf (old implementation) - per-client outgoing traffic limiter.
# Copyright (C) 2013 Edward Chernenko.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
###############################################################################

use strict;
use Data::Dumper;

# For what intervals of time (in seconds) do we need per-client statistics?
our @CHAINS = (10, 60, 300,
	3600, # 1 hour
	86400 # 1 day
);
our $ANALYZE_INTERVAL = 1; # seconds (integer)

open(F, "tcpdump -i eth0 -fn -v src port 80 -c 10000 |") || die $!;

my %P = (); # { client_ip1 => [ [ package1length, package1timestamp ], [ package2length, package2timestamp ], ... ] }
my %P24 = (); # { client_subnet1 => ..., client_subnet2 => } for /24 subnets
my %P16 = (); # for /16 subnets

my $last_analyze = time();
my $n = 1;
while(<F>)
{
#	print STDERR "\r$n..." if(!($n % 100));
	$n ++;

	my $time = time(); # Why bother parsing the timestamp? Difference less than 1 second is acceptable.
	my ($length, $ip) = ($_ =~ /length ([0-9]+).*> ([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)/);
	warn "IP not recognized, do you use IPv6? '$_'\n", next
 		unless($ip);
	
	$P{$ip} = [] unless(exists $P{$ip});
	
	do_register($ip, $length, $time);

	if($time - $last_analyze > $ANALYZE_INTERVAL)
	{
		$last_analyze = $time;
		do_analyze($time); # this won't take long, so we need not call time() again
	}
}
close F;
print STDERR "\r" . (" " x 40) . "\n";

sub do_register
{
	my($ip, $length, $time) = @_;
	
	my @i = split /\./, $ip;	
	my $ip16 = "$i[0].$i[1].0.0/16";
	my $ip24 = "$i[0].$i[1].$i[2].0/24";
	
	$P{$ip} = [] unless(exists $P{$ip});
	$P24{$ip24} = [] unless(exists $P24{$ip24});
	$P16{$ip16} = [] unless(exists $P16{$ip16});
	
	push @{$P{$ip}}, [ $length, $time ];
	push @{$P24{$ip24}}, [ $length, $time ];
	push @{$P16{$ip16}}, [ $length, $time ];
}

sub do_analyze
{
	my $time = shift;
	
	my $interval = 120;

	#
	# TEST function.
	#
	# For each client (or subnet - NYI) we calculate outgoing traffic to them
	# in the last $interval seconds
	#
	
	my %traffic_per_interval_and_ip = ();
	foreach my $ip(keys %P)
	{	
		my $traffic = 0;
		my $packets = $P{$ip}; 
	
		my $idx = $#$packets;
		
		# TODO: track all intervals from @CHAINS
		
		my $interval_start = $time - $interval;
		
		while($idx -- >= 0 && $packets->[$idx][1] >= $interval_start)
		{
			$traffic += $packets->[$idx][0];
		}
		$traffic_per_interval_and_ip{$interval}->{$ip} = $traffic;
	}
	
	print "\033[2J"; # clear the screen
	print "\033[0;0H"; # jump to 0,0

	print "Most traffic-consuming clients in $interval seconds:\n";
	my $h = $traffic_per_interval_and_ip{$interval};
	my @keys = sort { $h->{$b} - $h->{$a} } keys(%$h);
	
	@keys = @keys[0..19]
		if($#keys > 19);
	foreach my $c(@keys)
	{
		print $c . " | " .  ($h->{$c}/1024) . "Kb\n";
	}
	
	# BUG: traffic for some ips goes lower before the $interval from the beginning of the test
	
	print "----------------\n";
}

__END__






print "Most traffic-consuming clients in 10000 packets:\n";
@keys = sort { $TRAFFIC_PER_CLIENT{$b} - $TRAFFIC_PER_CLIENT{$a} } keys(%TRAFFIC_PER_CLIENT);
foreach $c(@keys)
{
	print $c . " | " .  ($TRAFFIC_PER_CLIENT{$c}/1024) . "Kb\n";
}

File: limittraf.c
/*
	limittraf - per-client outgoing traffic limiter.
	Copyright (C) 2013-2015 Edward Chernenko.

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
*/

/*
	Configuration options
*/

const char *ltCfgFile = "limittraf.conf";

const char *ltTcpDump = "tcpdump"; // "/usr/sbin/tcpdump";
const char *ltNetworkInterface = "em1"; // "eth0"
const char *ltTcpDumpOptions = "src port 80";

const char *ltTc = "tc"; // "/sbin/tc"

const char *ltWorkDir = "/tmp/limittraf";
const char *ltDbFile = "limittraf.db";
const char *ltLogFile = "limittraf.log";
const unsigned long ltLegitimateSearchEngineCacheExpires = 604800; // 604800 seconds = 1 week

const int ltAnalyzeInterval = 5;
const unsigned long ltMemoryDumpLevel = 10 * 1204 * 1024; // 10 megabytes

/*




*/

#include <stdio.h>
#include <pcre.h>
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <assert.h>

#include "conf.h"
#include "database.h"
#include "legsearch.h"
#include "actions.h"

/*
	Format of two lines printed by 'tcpdump -fnvKtq', with the newline removed from the first line:
		IP (tos 0x0, ttl 64, id 46394, offset 0, flags [DF], proto TCP (6), length 1492)    10.205.15.60.80 > 80.102.204.74.1155: tcp 1452
*/
static const char TcpDumpRequiredParams[] = "-fnvKtq"; /* These affect the format and therefore must be specified for parsing to success */

static const char TCPDUMP_REGEX[] = "length ([0-9]+).*> ([0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+)";
static const int TCPDUMP_LINE_MAX = 4096;

static pcre *tcpdump_regex;
static pcre_extra *tcpdump_extra;

time_t TIME = 0; /* = time(NULL), an approximation for timestamp of current packet */


/* ... */
static void Initialize(); /* Create the database and compile the regex */
static void Analyze(); /* Called every ltAnalyzeInterval seconds */

int main( void )
{
	char buffer[TCPDUMP_LINE_MAX];
	int ovector[9]; /* we have 2 values to match, +1 place for the entire regexp; PCRE requires 3x more space */
	char *TcpDumpCommand;
	FILE *tcpdump;
	int lineno;
	time_t last_analyze;
	int ret;
	
	/* IP and length are determined for each packet */
	char ip[17]; int length;
	char length_as_string[6]; /* MTU is never longer than 5 digits in decimal notation */
	
	Initialize();

#if 0
	const char *testhost =
 		"66.249.73.111"
		// "199.21.99.124"
		;
	fprintf(stderr, "%s is%s a legitimate search engine\n", testhost, is_legitimate_search_engine(testhost) ? "" : " NOT");
	fprintf(stderr, "%s is%s a legitimate search engine\n", testhost, is_legitimate_search_engine(testhost) ? "" : " NOT");
	fprintf(stderr, "%s is%s a legitimate search engine\n", testhost, is_legitimate_search_engine(testhost) ? "" : " NOT");
	exit(0);
#endif
	
	TcpDumpCommand = malloc(1024);
	if(!TcpDumpCommand)
	{
		fprintf(stderr, "malloc() for TcpDumpCommand failed: %s\n", strerror(errno));
		exit(1);
	}
	snprintf(TcpDumpCommand, 1024, "%s %s %s -i %s", ltTcpDump, ltTcpDumpOptions, TcpDumpRequiredParams, ltNetworkInterface);
	
	fprintf(stderr, "Starting %s\n", TcpDumpCommand);
	tcpdump = popen(TcpDumpCommand, "r");
	if(!tcpdump)
	{
		fprintf(stderr, "popen(%s) failed: %s\n", ltTcpDump, strerror(errno));
		return 1;
	}
	free(TcpDumpCommand);
	
	/* begin Main Loop */
	lineno = 0;
	last_analyze = time(NULL);
	while(fgets(buffer, TCPDUMP_LINE_MAX, tcpdump))
	{
		int off = strlen(buffer);
#ifndef NDEBUG
		if(!off)
		{
			fprintf(stderr, "tcpdump returned an empty line (bug in tcpdump or changed format). Exiting.\n");
			exit(1);
		}
#endif
		fgets(&buffer[off - 1], TCPDUMP_LINE_MAX - off, tcpdump);
		time(&TIME);

		if((++ lineno) % 100 == 0)
		{
			fprintf(stderr, "%i...\n", lineno);
			if(DbMemoryUsed() > ltMemoryDumpLevel)
			{
				CommitTransaction();
				CompactDb();
				BeginTransaction();
			}
		}
	
		ret = pcre_exec(tcpdump_regex, tcpdump_extra, buffer, strlen(buffer), 0, 0, ovector, 9);
		if(ret < 0)
		{
			fprintf(stderr, "pcre_exec() returned %i on [[%s]]\n", ret, buffer);
			continue;
		}

		pcre_copy_substring(buffer, ovector, 8, 1, length_as_string, 6);
		pcre_copy_substring(buffer, ovector, 8, 2, ip, 17);
		
//		fprintf(stderr, "[%6i] Length %s from %s\n", lineno, length_as_string, ip);
		
		length = atoi(length_as_string);
		Register(ip, length);

		if(TIME - last_analyze > ltAnalyzeInterval)
		{
			last_analyze = TIME;

			/* No error checking on commits because this is our private in-memory database */
			CommitTransaction();
			Analyze();
			BeginTransaction();
		}
	}
	
	return 0;
}

__attribute__((cold)) static void CompileTcpdumpRegex()
{
	const char *error; int erroffset;
	tcpdump_regex = pcre_compile(TCPDUMP_REGEX, PCRE_NO_UTF8_CHECK, &error, &erroffset, NULL);
	if(!tcpdump_regex)
	{
		fprintf(stderr, "Failed to compile regexp: '%s': %s\n", TCPDUMP_REGEX, error);
		exit(1);
	}
	tcpdump_extra = pcre_study(tcpdump_regex, PCRE_STUDY_JIT_COMPILE, &error); /* returned NULL is ok here */
}

__attribute__((cold)) static void Initialize()
{
	ReadConfiguration(ltCfgFile);
	
	if(mkdir(ltWorkDir, 0700) < 0 && errno != EEXIST)
	{
		fprintf(stderr, "mkdir(%s) failed: %s\n", ltWorkDir, strerror(errno));
		exit(1);
	}
	chdir(ltWorkDir);
	
	CompileTcpdumpRegex();
	InitializeLegSearch();
	InitializeActions();
	InitializeDb();
}
__attribute__((cold)) static void Terminate()
{
	TerminateDb();
	TerminateActions();
	TerminateLegSearch();
	if(tcpdump_extra) pcre_free_study(tcpdump_extra);
}

/**


*/
__attribute__((hot)) static void Analyze()
{
	fprintf(stderr, "Analyzing...\n");
	
	AnalyzeDb(); /* the actual work is performed here */
	CompactDb();
	LegSearch_Save();
}
